### ISSUE TITLE
Take into account dataInputs during mempool updateFamily

### ISSUE THREAD

**Issue by jasondavies** [ergoplatform/ergo#1156]
This is kind of useful for doing third-party CPFP by referencing an unconfirmed tx via data inputs.

**Comment by jasondavies** [comment]
However, make sure that each tx is only able to reference a "parent" unconfirmed tx once.

**Comment by jasondavies** [comment]
In fact, I think `updateFamily` currently has a bug where it updates the weight of a tx multiple times if a child transaction spends more than one output from a parent.

**Comment by kushti** [comment]
> In fact, I think `updateFamily` currently has a bug where it updates the weight of a tx multiple times if a child transaction spends more than one output from a parent.

@knizhnik can you take a look to this please ?

**Comment by jasondavies** [comment]
This comment doesn't really fully make sense either:

```
  /**
    * Form families of transactions: take in account relations between transaction when perform ordering.
    * If transaction X is spending output of transaction Y, then X weight should be greater than of Y.
    * Y should be proceeded prior to X or swapped out of mempool after X.
    * To achieve this goal we recursively add weight of new transaction to all transactions which
    * outputs it directly or indirectly spending.
    */
```

If `X` is spending `Y`, then we want `Y` to have a weight greater than `X`!

But I'm not sure that simply adding the weights together is correct.

Take the simple example of two transactions, `X` spending `Y`, both are unconfirmed.  The most advantageous ordering should take the combined fee and size to compute the weight, i.e. `X.weight = Y.weight = (X.fee + Y.fee) / (X.size + Y.size)`.  There would need to be a secondary ordering variable to ensure that `Y` always appears before `X`.

At the moment, it looks like it is just doing `Y.weight = X.fee/X.size + Y.fee/Y.size`.

**Comment by kushti** [comment]
Yes, averaging looks to be not proper. @knizhnik please fix this as well

**Comment by knizhnik** [comment]
> > In fact, I think `updateFamily` currently has a bug where it updates the weight of a tx multiple times if a child transaction spends more than one output from a parent.
> 
> @knizhnik can you take a look to this please ?

I wonder if it will be enough to change formula of weight recalculation for X spending Y to: Y.weight = Y.weigth + X.weight/size(X.inputs)

**Comment by knizhnik** [comment]
> But I'm not sure that simply adding the weights together is correct.

Well, I agree that correct formula of calculating cost of N transactions is sum(Ti.fee)/sum(Ti.size), not sum(Ti.fee/T.size). But our goal is not to calculate correct cost but provide optimal ordering. I am not sure how critical is to choose best order or "good enough" order is acceptable. Meaning of "optimal" also requires clarification. When we calculate our "balance"? After applying all N transactions? But new transaction constantly arrive so there is "final point"...
This is why I am not sure that your formula will work better than current one.
Assume three transactions X,Y,Z and and as assumed above X spending Y.
Let (X.weigth + Y.wiegth) > Z.weigth > X.wiegth > Y.weight
According to current formula we first take Y, then Z and then X.
But your formula will result different order: Y,X,Z which is worse.
